File: tourcount/src/main/java/com/wmstein/filechooser/AdvFileChooser.java
package com.wmstein.filechooser;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.ActivityInfo;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.view.KeyEvent;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;

import com.wmstein.tourcount.R;

import java.io.File;
import java.io.FileFilter;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * AdvFileChooser lets you select files from sdcard directory.
 * It will be called within WelcomeActivity and uses FileArrayAdapter and Option.
 * Based on android-file-chooser, 2011, Google Code Archiv, GNU GPL v3.
 * Modifications by wmstein on 18.06.2016
 */

public class AdvFileChooser extends Activity implements SharedPreferences.OnSharedPreferenceChangeListener
{
    private File currentDir;
    private FileArrayAdapter adapter;
    private FileFilter fileFilter;
    private ArrayList<String> extensions;
    private String filterFileName;
    private boolean screenOrientL; // option for screen orientation

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        prefs.registerOnSharedPreferenceChangeListener(this);
        screenOrientL = prefs.getBoolean("screen_Orientation", false);

        if (screenOrientL)
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
        } else
        {
            setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
        }

        setContentView(R.layout.list_view);

        Bundle extras = getIntent().getExtras();
        if (extras != null)
        {
            if (extras.getStringArrayList("filterFileExtension") != null)
            {
                extensions = extras.getStringArrayList("filterFileExtension");
                filterFileName = extras.getString("filterFileName");
                fileFilter = new FileFilter()
                {
                    @Override
                    public boolean accept(File pathname)
                    {
                        return
                            (
                                (pathname.getName().contains(".") &&
                                    pathname.getName().contains(filterFileName) &&
                                    extensions.contains(pathname.getName().substring(pathname.getName().lastIndexOf(".")))
                                )
                            );
                    }
                };
            }
        }

        // currentDir = new File ("/sdcard/")
        currentDir = new File(Environment.getExternalStorageDirectory().getPath());
        fill(currentDir);
    }

    // Back-key return from filechooser
    public boolean onKeyDown(int keyCode, KeyEvent event)
    {
        if (keyCode == KeyEvent.KEYCODE_BACK)
        {
            finish();
            return false;
        }
        return super.onKeyDown(keyCode, event);
    }

    // List only files in /sdcard
    private void fill(File f)
    {
        File[] dirs;

        if (fileFilter != null)
            dirs = f.listFiles(fileFilter);
        else
            dirs = f.listFiles();

        this.setTitle(getString(R.string.currentDir) + ": " + f.getName());
        List<Option> fls = new ArrayList<>();
        @SuppressLint("SimpleDateFormat") 
        DateFormat dform = new SimpleDateFormat("yyyy-MM-dd HH:mm");
        try
        {
            for (File ff : dirs)
            {
                if (!ff.isHidden())
                {
                    fls.add(new Option(ff.getName(), getString(R.string.fileSize) + ": "
                        + ff.length() + " B,  " + getString(R.string.date) + ": "
                        + dform.format(ff.lastModified()), ff.getAbsolutePath(), false, false, false));
                }
            }
        } catch (Exception e)
        {
            // do nothing
        }

        Collections.sort(fls);
        ListView listView = (ListView) findViewById(R.id.lvFiles);

        adapter = new FileArrayAdapter(listView.getContext(), R.layout.file_view, fls);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener()
        {
            @Override
            public void onItemClick(AdapterView<?> l, View v, int position,
                                    long id)
            {
                // TODO Auto-generated method stub
                Option o = adapter.getItem(position);
                assert o != null;
                if (!o.isBack())
                    doSelect(o);
                else
                {
                    currentDir = new File(o.getPath());
                    fill(currentDir);
                }
            }

        });
    }

    private void doSelect(final Option o)
    {
        //onFileClick(o);
        File fileSelected = new File(o.getPath());
        Intent intent = new Intent();
        intent.putExtra("fileSelected", fileSelected.getAbsolutePath());
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    @Override
    protected void onPause()
    {
        super.onPause();
    }

    @Override
    public void onStop()
    {
        super.onStop();
    }

    @Override
    protected void onDestroy()
    {
        super.onDestroy();
    }
    
    @Override
    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        screenOrientL = prefs.getBoolean("screen_Orientation", false);
    }

}

File: tourcount/src/main/java/com/wmstein/filechooser/AdvFileChooser.kt
package com.wmstein.filechooser

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.ListView
import android.widget.TextView
import com.wmstein.tourcount.R
import java.io.File
import java.io.FileFilter
import java.text.DateFormat
import java.text.SimpleDateFormat

/**
 * AdvFileChooser lets you select files from sdcard directory.
 * It will be called within WelcomeActivity and uses FileArrayAdapter and Option.
 * Based on android-file-chooser, 2011, Google Code Archiv, GNU GPL v3.
 * Adopted by wmstein on 2016-06-18,
 * last change in Java on 2022-05-21,
 * converted to Kotlin on 2023-07-09,
 * last edited on 2024-11-07.
 */
class AdvFileChooser : Activity() {
    private var currentDir: File? = null
    private var adapter: FileArrayAdapter? = null
    private var fileFilter: FileFilter? = null
    private var extension: String = ""
    private var filterFileName: String? = null

    @SuppressLint("SourceLockedOrientationActivity")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.list_view)

        val extras = intent.extras
        if (extras != null) {
            if (extras.getString("filterFileExtension") != null) {
                extension = extras.getString("filterFileExtension")!!
                filterFileName = extras.getString("filterFileName")
                fileFilter = FileFilter { pathname: File ->
                    pathname.name.contains(".") &&
                            pathname.name.contains(filterFileName!!) &&
                            extension.contains(
                                pathname.name.substring(
                                    pathname.name.lastIndexOf(
                                        "."
                                    )
                                )
                            )
                }
            }
        }

        // Set FileChooser Headline
        val fileHd = getString(R.string.fileHeadlineDB)
        val fileHead: TextView = findViewById(R.id.fileHead)
        fileHead.text = fileHd

        // currentDir = /storage/emulated/0/Documents/TourCount/
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) // Android 10+
        {
            currentDir = Environment.getExternalStorageDirectory()
            currentDir = File("$currentDir/Documents/TourCount")
        } else {
            currentDir = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS)
            currentDir = File("$currentDir/TourCount")
        }
        fill(currentDir!!)
    }

    // List only files in /sdcard
    private fun fill(f: File) {
        val dirs: Array<File>? = if (fileFilter != null) f.listFiles(fileFilter) else f.listFiles()
        this.title = getString(R.string.currentDir) + ": " + f.name
        val fls: MutableList<Option> = ArrayList()
        @SuppressLint("SimpleDateFormat") val dform: DateFormat =
            SimpleDateFormat("yyyy-MM-dd HH:mm")
        try {
            assert(dirs != null)
            if (dirs != null) {
                for (ff in dirs) {
                    if (!ff.isHidden) {
                        fls.add(
                            Option(
                                ff.name, getString(R.string.fileSize) + ": "
                                        + ff.length() + " B,  " + getString(R.string.date) + ": "
                                        + dform.format(ff.lastModified()), ff.absolutePath, false
                            )
                        )
                    }
                }
            }
        } catch (_: Exception) {
            // do nothing
        }
        fls.sort()
        val listView = findViewById<ListView>(R.id.lvFiles)
        adapter = FileArrayAdapter(listView.context, R.layout.file_view, fls)
        listView.adapter = adapter
        listView.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, position: Int, _: Long ->
                val opt = adapter!!.getItem(position)
                if (!opt.isBack) doSelect(opt) else {
                        currentDir = File(opt.path)
                        fill(currentDir!!)
                    }
            }
    }

    private fun doSelect(opt: Option?) {
        // onFileClick(opt);
        val fileSelected = File(opt!!.path)
        val intent = Intent()
        intent.putExtra("fileSelected", fileSelected.absolutePath)
        setResult(RESULT_OK, intent)
        finish()
    }

    public override fun onStop() {
        super.onStop()
    }

}
